apicast/src/configuration/service.lua

```diff
@@ -0,0 +1,167 @@
+local setmetatable = setmetatable
+local tostring = tostring
+local rawget = rawget
+local next = next
+local lower = string.lower
+local gsub = string.gsub
+local select = select
+local type = type
+
+local http_authorization = require 'resty.http_authorization'
+
+local _M = {
+
+}
+
+local mt = {
+  __index = _M
+}
+
+function _M.new(attributes)
+  return setmetatable(attributes or {}, mt)
+end
+
+local http_methods_with_body = {
+  POST = true,
+  PUT = true,
+  PATCH = true
+}
+
+local function read_body_args(...)
+  local method = ngx.req.get_method()
+
+  if not http_methods_with_body[method] then
+    return {}, 'not supported'
+  end
+
+  ngx.req.read_body()
+
+  local args = ngx.req.get_post_args()
+  local results = {}
+
+  for n=1, select('#', ...) do
+    results[n] = args[select(n, ...)]
+  end
+
+  return results
+end
+
+local function read_http_header(name)
+  local normalized = gsub(lower(name), '-', '_')
+  return ngx.var['http_' .. normalized]
+end
+
+local credentials_mt = {
+  -- nipairs = only non integer pairs
+  __pairs = function (t)
+    return function(_, k)
+      local v
+      repeat
+        k, v = next(t, k)
+      until k == nil or type(k) ~= 'number'
+      return k, v
+    end, t, nil
+  end
+}
+
+local backend_version_credentials = {
+  ['1'] = function(credentials)
+    local name = (credentials.user_key or 'user_key')
+    local user_key
+
+    if credentials.location == 'query' then
+      user_key = ngx.var['arg_' .. name] or read_body_args(name)[1]
+
+    elseif credentials.location == 'headers' then
+      user_key = read_http_header(name)
+
+    elseif credentials.location == 'authorization' then
+      local auth = http_authorization.new(ngx.var.http_authorization)
+
+      user_key = auth.userid or auth.password or auth.token
+    else
+      return nil, 'invalid credentials location'
+    end
+
+    return { user_key, user_key = user_key }
```

Comment:
because `table.concat` merges just integer keys, so if this object is meant to be used for both creating the cache key and as  a hash to make a query string from it needs to be this weird mixture of array/hash